tests/src/test/java/org/eclipse/hono/tests/mqtt/MqttConnectionIT.java

```diff
@@ -588,4 +588,85 @@ public void testConnectX509FailsForDisabledTenant(final VertxTestContext ctx) {
             ctx.completeNow();
         }));
     }
+
+    /**
+     * Verifies that the adapter closes the connection to a device when the registration data of the device
+     * is deleted.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void testDeviceConnectionIsClosedOnDeviceDeleted(final VertxTestContext ctx) {
+        final Tenant tenant = new Tenant();
+
+        final Promise<Void> connectionClosedPromise = Promise.promise();
+        // GIVEN a connected device
+        helper.registry
+                .addDeviceForTenant(tenantId, tenant, deviceId, password)
+                .compose(ok -> connectToAdapter(
+                        IntegrationTestSupport.getUsername(deviceId, tenantId), password))
+                .compose(conAckMsg -> {
+                    ctx.verify(() -> assertThat(conAckMsg.code()).isEqualTo(MqttConnectReturnCode.CONNECTION_ACCEPTED));
+                    mqttClient.closeHandler(remoteClose -> connectionClosedPromise.complete());
+                    // WHEN device data is removed
+                    return helper.registry.deregisterDevice(tenantId, deviceId);
+                })
+                // THEN the device connection is closed
+                .compose(ok -> connectionClosedPromise.future())
+                .onComplete(ctx.succeedingThenComplete());
+    }
+
+    /**
+     * Verifies that the adapter closes the connection to a device when the tenant of the device
+     * is deleted.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void testDeviceConnectionIsClosedOnTenantDeleted(final VertxTestContext ctx) {
+        final Tenant tenant = new Tenant();
+
+        final Promise<Void> connectionClosedPromise = Promise.promise();
+        // GIVEN a connected device
+        helper.registry
+                .addDeviceForTenant(tenantId, tenant, deviceId, password)
+                .compose(ok -> connectToAdapter(
+                        IntegrationTestSupport.getUsername(deviceId, tenantId), password))
+                .compose(conAckMsg -> {
+                    ctx.verify(() -> assertThat(conAckMsg.code()).isEqualTo(MqttConnectReturnCode.CONNECTION_ACCEPTED));
+                    mqttClient.closeHandler(remoteClose -> connectionClosedPromise.complete());
+                    // WHEN tenant data is removed
+                    return helper.registry.removeTenant(tenantId);
+                })
+                // THEN the device connection is closed
+                .compose(ok -> connectionClosedPromise.future())
+                .onComplete(ctx.succeedingThenComplete());
+    }
+
+    /**
+     * Verifies that the adapter closes the connection to a device when registration data for all devices of the device
+     * tenant is deleted.
+     *
+     * @param ctx The vert.x test context.
+     */
+    @Test
+    public void testDeviceConnectionIsClosedOnAllDevicesOfTenantDeleted(final VertxTestContext ctx) {
```

Comment:
how about reducing the code duplication by creating a private method accepting a Supplier<Future<?>> which you can use to trigger the device/tenant deletion?